psi4/driver/procrouting/response/scf_products.py

```diff
@@ -0,0 +1,601 @@
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This file is part of Psi4.
+#
+# Psi4 is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# Psi4 is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+
+from itertools import product
+
+import numpy as np
+
+from psi4 import core
+from psi4.driver.p4util.exceptions import *
+
+
+"""
+- H1/ H2 docstrings to keep them straight
+- RHF equations
+
+Aia,jb  = (E_a - E_i) +     J    -    K
+        = (E_a - E_i) + 2(ia|jb) - (ij|ab)
+
+Bia,jb  =     J    -   K^T
+        = 2(ai|bj) - (aj|bi)
+
+H2ia,jb =                A                   -           B
+        = [(E_a - E_i) + 2   J    -    K   ]  - [2  J     -    K^T ]
+        = [(E_a - E_i) + 2(ia|jb) - (ij|ab)]  - [2 (ai|bj)  - (aj|bi)]
+        = (E_a - E_i) -    K    +    K^T
+        = (E_a - E_i) - (ij|ab) + (aj|bi)
+
+H1ia,jb =              A                    +       B
+        = [(E_a - E_i) + 2   J    -    K   ] + [2 J        -   K^T  ]
+        = [(E_a - E_i) + 2(ia|jb) - (ij|ab)] + [2 (ai|bj)  - (aj|bi)]
+        = [(E_a - E_i) + 4   J    -    K     -   K^T]
+        = [(E_a - E_i) + 4(ia|jb) - (ij|ab)  - (aj|bi)]
+
+"""
+
+
+class SingleMatPerVector:
+    """Matrix operations for RHF-like systems
+    """
+
+    @staticmethod
+    def vector_dot(X, Y):
+        return X.vector_dot(Y)
+
+    @staticmethod
+    def vector_scale(a, X):
+        X.scale(a)
+        return X
+
+    @staticmethod
+    def vector_axpy(a, X, Y):
+        Y.axpy(a, X)
+        return Y
+
+    @staticmethod
+    def vector_copy(X):
+        return X.clone()
+
+    @staticmethod
+    def vector_transpose(X):
+        return X.transpose()
+
+
+class PairedMatPerVector:
+    """Matrix operations for UHF-like systems
+    """
+
+    @staticmethod
+    def vector_dot(X, Y):
+        dot = X[0].vector_dot(Y[0])
+        dot += X[1].vector_dot(Y[1])
+        return dot
+
+    @staticmethod
+    def vector_scale(a, X):
+        X[0].scale(a)
+        X[1].scale(a)
+        return X
+
+    @staticmethod
+    def vector_axpy(a, X, Y):
+        Y[0].axpy(a, X[0])
+        Y[1].axpy(a, X[1])
+        return Y
+
+    @staticmethod
+    def vector_copy(X):
+        return [X[0].clone(), X[1].clone()]
+
+    @staticmethod
+    def vector_transpose(X):
+        return [X[0].transpose(), X[1].transpose()]
+
+
+class ProductCache:
+    """Caches product vectors
+    """
+
```

Comment:
Any advantage in using `__slots__ = ['_products']` here?

Reply:
While adding `__slots__` wouldn't change how this behaves at all I don't expect that would make much of a difference as there will only be one of these things per product engine, and only one product engine at a given time.